Feat/fancybox set

kooropatfa opened the pull request:
Bases on https://github.com/Madek/madek-graphql-api/pull/7
#closes https://github.com/Madek/madek-graphql-api/issues/3

lotcod3 committed:
Reorganized specs in more reasonable way to reflect code structure better

lotcod3 committed:
Added setQuery

lotcod3 committed:
Updated schema.graphql

lotcod3 committed:
Updated and reorganized specs

lotcod3 committed:
Simplfied response_as_hash method

eins78 committed:
update datalayer

eins78 committed:
scripts

eins78 committed:
scripts

eins78 committed:
load fewer frameworks from rails

eins78 committed:
spec: ensure that IntrospectionQuery works

otherwise GraphiQL UI does not work

eins78 committed:
fix: class-based enum definitions

when using `define` the IntrospectionQuery fails with following error:

Invariant: can't introspect non-class-based type

eins78 commented:
@kooropatfa I tried to use graphiql but the "IntrospectionQuery" (used to get the Schema from the API instead of loading a file) was broken, because of the way the Enums were defined (I dont quite understand why but thats what the `graphql` gem told me).

So, I added a new spec to `master` to make sure that this specific query wont break in the future: https://github.com/Madek/madek-graphql-api/blob/c6e065cc0503e3dd953470aa714126e11409e48c/spec/graphql/queries/introspection.rb

I also merged in `master` into this branch and added a minimal fix for the enums so the web ui works again.

eins78 commented on app/graphql/madek_graphql_schema.rb:
```diff
@@ -6,32 +8,31 @@ class MadekGraphqlSchema < GraphQL::Schema
-    name 'OrderByEnum'
+  # NOTE: minimally adapted so that introspection works,
+  #       docs say it should be separated and nested in Types::,
+  #       see https://graphql-ruby.org/type_definitions/enums.html#defining-enum-types
```
@kooropatfa this note is for you ;)

eins78 commented:
@kooropatfa I also updated the datalayer (still in flux because of the ongoing ruby upgrade, dont worry about it for now) and ran the spec that you said works locally but not in CI.
For me, it also fails locally, with the same error, so i suspect there is something broken/dirty in your DB/env/etc. Please run `./scripts/dev-reset` and try again.

eins78 commented on spec/graphql/queries/set_query_spec.rb:
```diff
@@ -0,0 +1,69 @@
+describe 'getCollection' do
+  before(:all) do
```
```suggestion
  before(:each) do
```

Always use `each` when creating data, so each **example** is independent.
In the "getCollection loads collections by ID" example the count was wrong because you made 1 invisible in "returns collection with only publicly visible media entries".

michalpodlecki commented on spec/graphql/queries/set_query_spec.rb:
```diff
@@ -0,0 +1,65 @@
+describe 'getCollection' do
+  before(:each) do
+    FactoryGirl.build(:meta_datum_title_with_collection).save(validate: false)
+    # When I try to just create :meta_datum_title_with_collection I get
```
@kooropatfa to fix this issue add `config.active_record.belongs_to_required_by_default = false` to `config/application.rb` after `load_defaults` call

lotcod3 committed:
fix specs

lotcod3 committed:
update schema

lotcod3 committed:
Fix bad merge

lotcod3 committed:
Merge branch 'master' into feat/fancybox_set

eins78 commented:
@kooropatfa when testing your branch in graphiql, try the example query from the doc. You can use factories or just restore `personas.pgbin` from the datalayer submodule to get some test data.

eins78 commented:
@kooropatfa I adapted the example query to use connections. 

small errors:
* `metaData.value` should be `values` and always return an array (because there are types than can have multiple values)
* the `mediaTypes` filter on `MediaEntryConnection` as well as `PreviewConnection` should take an array (so one could get audios AND videos)

bugs:
*  using `mediaTypes` filter on `PreviewConnection` throws error "wrong number of arguments (given 1, expected 0)"
* when giving a non-exisiting id to `set`/`Collection`, it throws error "uninitialized constant Collection::CustomUrl". this comes from some copy/pasted models, there is a fallback in case the ID is not found that should be removed.
    ```graphql
    { set(id: "4C35291B-7522-43CE-A562-34EAE1BF797A") { id } }
    ```


```graphql
{
  set(id: "e9f34e2c-b844-4297-b701-7e512c65e3b5") {
    id
    url
    metaData {
      nodes {
        id
        metaKey {
          id
        }
        # `values` is always plural/array
        value {
          string
        }
      }
    }
    # filter `mediaTypes` should take array!
    childMediaEntries(mediaTypes: IMAGE) {
      nodes {
        id
        url
        metaData {
          nodes {
            id
            metaKey {
              id
            }
            value {
              string
            }
          }
        }
        # * filter `mediaTypes` should take array!
        # * when using filter `mediaTypes` it crashes!
        mediaFile {
          previews {
            nodes {
              id
              url
              contentType
              mediaType
              sizeClass
            }
          }
        }
      }
    }
  }
}
```

eins78 reviewed (changes requested):
basically works, very nice!

please adress the smaller bugs before merging: https://github.com/Madek/madek-graphql-api/pull/8#issuecomment-507180899

lotcod3 committed:
Fix changes requested in #8 (comment):

- change metaData.value to values array
- fix mediaTypes filter on previewConnection
- change mediaType filter on mediaEntryConnection and previewConnection
to take an array of types as argument
- added MadekErrors class for easier errors handling
- named enums more precisely and separated their definitions from schema

kooropatfa commented:
Fixed query with mediaType filters on correct fields  (mediaEntries and previews)

`{
  set(id: "d0f3c560-26e1-4089-b6b9-0cfb3eca62fb") {
    id
    url
    metaData {
      nodes {
        id
        metaKey {
          id
        }
        # `values` is always plural/array
        values {
          string
        }
      }
    }
    # filter `mediaTypes` should take array!
    childMediaEntries(mediaTypes: [IMAGE, AUDIO]) {
      nodes {
        id
        url
        metaData {
          nodes {
            id
            metaKey {
              id
            }
            values {
              string
            }
          }
        }
        # * filter `mediaTypes` should take array!
        # * when using filter `mediaTypes` it crashes!
        mediaFile {
          previews(mediaTypes: [IMAGE, AUDIO]) {
            nodes {
              id
              url
              contentType
              mediaType
              sizeClass
            }
          }
        }
      }
    }
  }
}`
